source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -5618,6 +5664,31 @@ export class ColumnSet extends StylableCardElementContainer {
     private _columns: Array<Column> = [];
     private _renderedColumns: Array<Column>;
 
+    private createColumnInstance(json: any, errors: Array<HostConfig.IValidationError>): Column {
+        return createCardObjectInstance<Column>(
+            this,
+            json,
+            [], // Forbidden types not supported for elements for now
+            (typeName: string) => {
+                return !typeName || typeName === "Column" ? new Column() : null;
+            },
+            (typeName: string, errorType: InstanceCreationErrorType) => {
+                if (errorType == InstanceCreationErrorType.UnknownType) {
+                    return {
+                        error: Enums.ValidationError.UnknownElementType,
+                        message: "Invalid element type " + typeName + " in a ColumnSet. Attempting to fall back."
+                    }
+                }
+                else {
+                    return {
+                        error: Enums.ValidationError.ElementTypeNotAllowed,
+                        message: "Element type " + typeName + " is not allowed in this context."
```

Comment:
Hmm we are already generalizing the concept of "invalid types" - why not consolidate this into a single handler for forbidden types? Why the special handling of UnknownType and ElementTypeNotAllowed?

Reply:
If you only get "invalid element", do you know why it's invalid? Is it because its type is unknown? Is it because you're using an element in a context where it can't be used? Is it for another (future) reason?

Say you're coding and you write this statement: `a = b`. The compiler says "Invalid statement"; would it be better if the compiler said "Can't assign b to a because a is a Boolean and b is a string"

Reply:
Also, please remember that the JS renderer supports the designer. The designer needs to be as helpful as possible.

Reply:
I guess i am having a hard time distinguishing between "INvalid element type <> in a columnset." vs "Element type not allowed in this context" -- what is invalid and what is not allowed?

---
In reply to: [279437943](https://github.com/Microsoft/AdaptiveCards/pull/2771#discussion_r279437943) [](ancestors = 279437943)

Reply:
Sorry, I should have looked at the code more carefully, which would have helped me better understand what you were saying. The messages are incorrect indeed (I copy pasted and forgot to cleanup). I'll push a change. The final messages will be:
- Unknown element type <element type>. Attempting to fall back.
- Element type <element type> isn't allowed in a ColumnSet.